Channel Branding Feature & Various Fixes/Improvements & Firefox chromium merged development

namakeingo opened the pull request:
# Disclaimer

This is a huge PR, and I am sorry about that. I totally understand it will take a while to review. I can answer any question you might have, just let me know.
The CIs have additional secrets that will need to be added to your Repo Secrets. If you could give me Contributor rights, I can add them. 
I updated the documentation and included comments, but happy to do any other changes you think appropriate.

## New Feature

- I added the feature to untranslate the channel header and about section. 
_EDIT: It defaults to YouTube Internal API with original language detection using i18n. Optionally the user can provide a API Key to use_ the YouTube Data API, so I added an additional section in the `popup.html` and the related `options.js` changes to handle the section. I also added instructions for the users on how to obtain a YouTube Data API Key [YOUTUBE_DATA_API_KEY.md](https://github.com/namakeingo/yt-anti-translate-firefox/blob/firefox-chromium-merged-development/YOUTUBE_DATA_API_KEY.md)
The code for this feature is contained in [content_channelbranding.js](https://github.com/namakeingo/yt-anti-translate-firefox/blob/firefox-chromium-merged-development/app/src/content_channelbranding.js) and is loaded as a `content_script` rather than injected as it needs access to `chrome.storage.sync.get` to use the API Key _EDIT: or `(browser||chrome).i18n` to detect the original language_.
- The extension is now also untranslating the highlighted videos on a Channel Page. Both the titles and the description
![image](https://github.com/user-attachments/assets/f249825d-f2d3-44be-9ba5-ef153b41cab1)
- The extension can also now handle the video links that appear on the player when in full screen ( https://github.com/zpix1/yt-anti-translate/issues/25#issuecomment-2788374298 ) and also the link to the most recent video of the same YouTube creator that is in the short player.
![image](https://github.com/user-attachments/assets/b2c2f937-8ffc-4350-b840-52cd37631ace)
- The extension is also untranslating the channel name above the description of a video (example in Thai cause MrBeast is MrBeast for russian translation)
![image](https://github.com/user-attachments/assets/813c020e-b480-4d63-b94f-1d41a4e71902)

## Improvements

- I added a [global.js](https://github.com/namakeingo/yt-anti-translate-firefox/blob/firefox-chromium-merged-development/app/src/global.js) file using a frozen Window DOM property where we can define all the reusable code to avoid repetition. The file is loaded as a `content_script` with `"run_at": "document_end"` so that it exists earlier than other scripts (default is "document_idle"). `global.js` is also injected into the page for use by other injected scripts with [content_injectglobal.js](https://github.com/namakeingo/yt-anti-translate-firefox/blob/firefox-chromium-merged-development/app/src/content_injectglobal.js) that is also set as  `"run_at": "document_end"`.
- I added functions to check that Nodes are computed as visible. As YouTube uses a components approach, no element is really ever removed unless a reload happens. Therefore, elements still exist even when completely invisible, so the extension is doing work even if no longer needed. ( will help with #40 )
- Similarly to the last, I added checks for the elements to be inside the Viewport so that we only do work on elements that the user can actually see. The viewport is currently extended by 50% to allow preload (that value is set in `global.js` as `VIEWPORT_EXTENSION_PERCENTAGE_FRACTION`, we could consider exposing that as a user option, but it would be a later thing). To handle elements that can enter the Viewport (otherVideos and otherShorts) I also added `IntersectionObserver` for both that trigger the `untranslateOtherVideos` and `untranslateOtherShortsVideos` strictly for the elements detected as newly intersecting by the Observer. ( will help with #40 )

## Fixes

- Fixed bug with audio (dubbing) not always untranslating due to not using the boolean result of `setAudioTrack` and the value `player.unstranslatedAudio` persisting even when going to the next video or short.
- Fixed the document.title untranslation not working when the spacing was inconsistent (normalise before replace).
- Fix duplication of untranslated fake node (on load, multiple hidden elements were being created before settling)
- Fixed #22

# CIs / Testing

- Updated Playwright configs to run tests for both Chromium and Firefox and use different locales where appropriate.
- Updated the Playwright CI to handle the above and also to upload screenshots as artifacts.
- Playwright now does authentication to a Google test account I created just for this purpose. #35 
It's not that YouTube was not working when the browser was headless or controlled, but it was enforcing a login to verify that the user is not a bot. Logging in solves the issue. The login is handled by a helper function on the setup of any test, and StorageState is saved and loaded to reuse the auth cookies to avoid issues with logins being too frequent. This needs the secrets to work properly.
- Added a helper to handle distributions of test files where user options can be changed before the test run, so that we can also test scenarios with different options from the default.
- Added a helper to download and enable the uBlock extension during tests, to prevent tests from failing because of Ads. Works perfectly in Firefox, not so perfectly in Chromium cause uBlock Lite functionality is limited.
- Added additional Playwright tests, both for existing features and the features that I added.
- Created a new CI [release.yml](https://github.com/namakeingo/yt-anti-translate-firefox/blob/firefox-chromium-merged-development/.github/workflows/release.yml) that uses the `CHANGELOG.md` to automatically create a new tagged Release when a PR to main is merged and triggers the publish CIs.
- Renamed `publish.yml` to `publish-chromium.yml` and created a new CI [publish-firefox.ym](https://github.com/namakeingo/yt-anti-translate-firefox/blob/firefox-chromium-merged-development/.github/workflows/publish-firefox.yml). Currently configured to run only when the repo owner is namakeingo, so that it will be skipped in your repo.

zpix1 commented on CHANGELOG.md:
```diff
@@ -5,49 +5,78 @@ All notable changes to this project will be documented in this file.
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
-## [1.17.0] - 05.04.2025
+## [1.18.1.6] - 2025-05-27
```
not a big fun of 1.1.1.1 versions, can we use just 3 numbers?

zpix1 commented on README.md:
```diff
@@ -1,23 +1,60 @@
 
+### Extra Features
+
+Extra features need a YouTube Data API Key to be populated in the extension settings. [Read more on how to obtain](https://github.com/namakeingo/yt-anti-translate-firefox/blob/main/YOUTUBE_DATA_API_KEY.md)
```
update url to current repo then?

zpix1 commented on app/pages/popup.html:
```diff
@@ -102,38 +134,64 @@
+        </div>
+        <div class="option" id="api-key-container">
+          <div class="small">
+            <p>Additional features require a <b>YouTube Data API Key</b></p>
```
move link https://github.com/zpix1/yt-anti-translate/blob/main/YOUTUBE_DATA_API_KEY.md under here (rm b)

namakeingo replied:
link moved

zpix1 commented on app/pages/popup.html:
```diff
@@ -102,38 +134,64 @@
-        <span>Reload page automatically</span>
-        <input id="reload-checkbox" type="checkbox" checked />
+      <div class="footer">
+        <a target="_blank" href="https://chrome.google.com/webstore/detail/yt-anti-translate/ndpmhjnlfkgfalaieeneneenijondgag">Rate Chrome extension</a> • 
```
too much links, must be at most 3 (check if current browser is firefox/chrome). If you still want donations to be displayed, make donation link direct to not to my site, but to github support section in README.

BTW, no donations even after 30k users, so maybe rm that completly

zpix1 commented on app/src/options.js:
```diff
@@ -67,11 +85,60 @@ function checkboxUpdate() {
+          youtubeDataApiKey: newApiKey
+        },
+        () => {
+          console.log("API key saved:", newApiKey);
```
better not log api keys

zpix1 commented on tests/helpers/ExtensionsFilesHelper.ts:
```diff
@@ -0,0 +1,194 @@
+    }
+  };
+
+  // Create the testDist directory if it doesn't exist
```
too much slop comments

namakeingo replied:
not sure what you mean

zpix1 replied:
I mean some comments are too verbose

zpix1 commented on tests/helpers/ExtensionsFilesHelper.ts:
```diff
@@ -0,0 +1,194 @@
+import fs from 'fs';
```
ublock origin still work in chrome?

namakeingo replied:
there is a Chrome version for mv3. It's not great but still somewhat works. Still helps with tests as some ads expect user interaction to skip

zpix1 reviewed (changes requested):
There a lot of changes really...

About UI, I think changes are really required. For now settings view is too ugly.

About global.js, idk, maybe for now - ok. Actually, its better to have something like typescript + bundler for that stuff when extension becomes larger. 

Thanks much for your work 👍

namakeingo commented on .github/workflows/publish-firefox.yml:
```diff
@@ -0,0 +1,107 @@
+name: Update Firefox Extension
```
if you can give me contributor role, sure I can drop the forl

zpix1 commented:
For youtube API key, it is possible for me to setup a server which obtains original stuff (while not sure chrome web store will allow that). I think obtaining youtube api key may be too hard for a regular user.

zpix1 commented on CHANGELOG.md:
```diff
@@ -5,49 +5,78 @@ All notable changes to this project will be documented in this file.
+
+### Changed
+
+- Optimise start-up logic for better performance
```
no need, thanks

namakeingo commented on app/pages/popup.html:
```diff
@@ -102,38 +134,64 @@
+          <span>Reload page automatically</span>
+          <input id="reload-checkbox" type="checkbox" checked />
+        </div>
+        <div class="option" id="api-key-container">
```
![image](https://github.com/user-attachments/assets/554101bf-926a-447b-b122-397e8e3d7a28)

namakeingo commented on app/pages/popup.html:
```diff
@@ -102,38 +134,64 @@
+        <div class="option" id="api-key-container">
+          <div class="small">
+            <p>Additional features require a <b>YouTube Data API Key</b></p>
+            <p><a target="_blank" href="https://github.com/zpix1/yt-anti-translate/blob/main/YOUTUBE_DATA_API_KEY.md">How do I obtain an API Key?</a></p>
```
done

namakeingo commented:
> Also, you sure there is no way to get branding without youtube API key?

Unfortunately, I could not find a better alternative at the moment. I tried digging in their js files to no avail

> About global.js, idk, maybe for now - ok. Actually, its better to have something like typescript + bundler for that stuff when extension becomes larger.

Yeah typescript would be better, but for the time being, I think `global.js` is already a huge help for reusable stuff

> About UI, I think changes are really required. For now settings view is too ugly.

Hopefully better now :)

> Also I plan to carefully review untranslation logic and tests a bit later.

I extended the tests to avoid breaking stuff, but do let me know if something still broke

> For youtube API key, it is possible for me to setup a server which obtains original stuff (while not sure chrome web store will allow that). I think obtaining youtube api key may be too hard for a regular user.

I am guessing that would have costs involved. I think with the guide getting an API Key is not too difficult.

zpix1 commented:
Very nice, you may try this JS (for me, it works for channel ids of MrBeast `UCX6OQ3DkcsbYNE6H8uQQuVA` and Ryan Trahan `UCnmGIkw-KdI0W5siakKPKog`)
```
/**
 * Fetch the About/branding section of a YouTube channel.
 *
 * @param {string} ucid   Channel ID (starts with “UC…”)
 * @param {string} locale Optional BCP-47 tag, e.g. "it-IT" or "fr"
 *                        Defaults to the user’s browser language.
 * @returns {object}      The full browse() JSON for further use.
 */
async function getChannelBranding(ucid, locale = navigator.language) {
  // ── 1. split the locale into hl (= language) and gl (= region) ──
  //    • “it-IT” → ["it","IT"]
  //    • “fr”    → ["fr"]
  const [hl, region] = locale.split(/[-_]/);           // tolerate "pt-BR" or "pt_BR"
  const gl = region || hl.toUpperCase();               // fallback: "fr" → "FR"

  // ── 2. build the request body ──
  const body = {
    context: {
      client: {
        hl,                    // lingua
        gl,                    // paese
        clientName:  "WEB",
        clientVersion: "2.20250527.00.00"
      }
    },
    browseId: ucid,
    // “about” tab protobuf — old first, then fall back to the newer encoding
    params: "EgVhYm91dA=="      // {2:string:"about"} :contentReference[oaicite:0]{index=0}
  };

  // ── 3. fire the request ──
  const ep = "https://www.youtube.com/youtubei/v1/browse?prettyPrint=false";
  let res = await fetch(ep, {method: "POST", headers: {"content-type":"application/json"}, body: JSON.stringify(body)});

  // YT sometimes rejects the legacy protobuf with 400; try the newer one.
  if (res.status === 400) {
    body.params = "EgVhYm91dPIGBAoCEgA=";              // wrapped “about” protobuf (2023-present) :contentReference[oaicite:1]{index=1}
    res = await fetch(ep, {method: "POST", headers: {"content-type":"application/json"}, body: JSON.stringify(body)});
  }

  const json = await res.json();

  // ── 4. quick-and-dirty console demo ──
  const hdr = json.header?.c4TabbedHeaderRenderer
           || json.header?.interactiveTabbedHeaderRenderer;

  console.log({
    banner : hdr?.banner?.thumbnails?.at(-1)?.url,      // full-width channel art
    avatar : hdr?.avatar?.thumbnails?.at(-1)?.url,      // profile picture
    title  : hdr?.title?.simpleText,                    // channel name
    subs   : hdr?.subscriberCountText?.simpleText       // “182 M subscribers”
  });

  return json;  // hand back the full response for further digging
}

// EXAMPLE ─ get branding in Italian
getChannelBranding("UCczv-fUYkJH09vA4ZuCQ", "it-IT");
```

namakeingo commented:
That was great help, thanks. I changed it to:
```
/**
 * Fetch the About/branding section of a YouTube channel.
 *
 * @param {string} ucid   Channel ID (starts with “UC…”)
 * @param {string} locale Optional BCP-47 tag, e.g. "it-IT" or "fr". Defaults to the user’s browser language.
 * @returns {object}      The title and description branding.
 */
async function getChannelBranding(ucid, locale = navigator.language) {
  // ── 1. split the locale into hl (= language) and gl (= region) ──
  //    • “it-IT” → ["it","IT"]
  //    • “fr”    → ["fr"]
  const [hl, region] = locale.split(/[-_]/);           // tolerate "pt-BR" or "pt_BR"
  const gl = region || hl.toUpperCase();               // fallback: "fr" → "FR"

  // ── 2. build the request body ──
  const body = {
    context: {
      client: {
        hl,                    // language
        gl,                    // country
        clientName: "WEB",
        clientVersion: "2.20250527.00.00"
      }
    },
    browseId: ucid,
    // “about” tab protobuf — old first, then fall back to the newer encoding
    params: "EgVhYm91dA=="      // {2:string:"about"} :contentReference[oaicite:0]{index=0}
  };

  const requestIdentifier = `${body.context.hl}-${body.context.gl}_${body.browseId}`

  // Check cache
  if (window.YoutubeAntiTranslate.cache.has(requestIdentifier)) {
    return window.YoutubeAntiTranslate.cache.get(requestIdentifier);
  }

  // ── 3. fire the request ──
  const ep = "https://www.youtube.com/youtubei/v1/browse?prettyPrint=false";
  let res = await fetch(ep, { method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify(body) });

  // YT sometimes rejects the legacy protobuf with 400; try the newer one.
  if (res.status === 400) {
    body.params = "EgVhYm91dPIGBAoCEgA=";              // wrapped “about” protobuf (2023-present) :contentReference[oaicite:1]{index=1}
    res = await fetch(ep, { method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify(body) });
  }

  const json = await res.json();

  // ── 4. quick-and-dirty console demo about ──
  const metadata = json.metadata?.channelMetadataRenderer;

  console.log({
    title: metadata?.title, // channel name
    fullAbout: metadata?.description // truncated about
  });

  const result = {
    title: branding.title,
    description: branding.description
  };

  // Store in cache
  window.YoutubeAntiTranslate.cache.set(requestIdentifier, result);

  return result;
}

// EXAMPLE ─ get branding in Italian for MrBeast
getChannelBranding("UCX6OQ3DkcsbYNE6H8uQQuVA", "it-IT");
// EXAMPLE ─ get branding in Italian for Cartoni Morti
getChannelBranding("UChn336v-fUYkJH09vA4ZuCA", "it-IT");
```

Like this, it is **almost** a drop-in replacement for my current function `async function fetchChannelTitleAndDescription(youtubeDataApiKey)`
However, the channel ID (UC...) is not always available on the page. There is a `var = ytInitialData` as well as some hidden <link> elements with `rel` "canonical" or "alternate". But unfortunately, those are not kept updated when you select another channel without reloading (for example, if you pick one of the other channels by MrBeast)

Also, do you have any idea how we could find the original language for the channel?

namakeingo commented:
It is getting close, but it is useless unless we can tell what the correct language to use is for the current channel.

```javascript
/**
 * Retrieve the UCID of a channel using youtubei/v1/search
 * @param {string} query the YouTube channel handle (e.g. "@mrbeast" or "MrBeast")
 * @returns {string} channel UCID
 */
async function lookupChannelId(query) {
  // build the request body ──
  const body = {
    context: {
      client: {
        clientName: "WEB",
        clientVersion: "2.20250527.00.00"
      }
    },
    query,
    // "EgIQAg==" = filter=channels  (protobuf: {12: {1:2}})
    params: "EgIQAg=="
  };

  const requestIdentifier = `youtubei/v1/search_${body}`

  // Check cache
  if (window.YoutubeAntiTranslate.cache.has(requestIdentifier)) {
    return window.YoutubeAntiTranslate.cache.get(requestIdentifier);
  }

  const r = await fetch("https://www.youtube.com/youtubei/v1/search?prettyPrint=false", {
    method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify(body)
  }).then(r => r.json());

  const channelUcid = r.contents?.twoColumnSearchResultsRenderer
    ?.primaryContents.sectionListRenderer
    ?.contents[0].itemSectionRenderer.contents[0]
    ?.channelRenderer?.channelId || null;

  // Store in cache
  window.YoutubeAntiTranslate.cache.set(requestIdentifier, result);

  return channelUcid;
}

/**
 * Retrieved the Channel UCID (UC...) for the current Channel page using window.location and lookupChannelId() search
 * @returns {string} channel UCID
 */
async function getChannelUCID() {
  if (window.location.pathname.startsWith("/channel/")) {
    var match = window.location.pathname.match(/\/channel\/([\w-]+)/);
    return match ? `${match[1]}` : null;
  }

  var handle = null;
  if (window.location.pathname.startsWith("/c/")) {
    var match = window.location.pathname.match(/\/c\/([\w-]+)/);
    handle = match ? `${match[1]}` : null;
  }
  else if (window.location.pathname.startsWith("/@")) {
    var match = window.location.pathname.match(/\/(@[\w-]+)/);
    handle = match ? `${match[1]}` : null;
  }
  else if (window.location.pathname.startsWith("/user/")) {
    var match = window.location.pathname.match(/\/user\/(@[\w-]+)/);
    handle = match ? `${match[1]}` : null;
  }
  return await lookupChannelId(handle);
}

/**
 * Fetch the About/branding section of a YouTube channel.
 * @param {string} ucid   Otional Channel ID (starts with “UC…”). Defaults to UCID of the current channel
 * @param {string} locale Optional BCP-47 tag, e.g. "it-IT" or "fr". Defaults to the user’s browser language.
 * @returns {object}      The title and description branding.
 */
async function getChannelBranding(ucid = null, locale = navigator.language) {
  // split the locale into hl (= language) and gl (= region)
  //    • “it-IT” → ["it","IT"]
  //    • “fr”    → ["fr"]
  const [hl, region] = locale.split(/[-_]/);           // tolerate "pt-BR" or "pt_BR"
  const gl = region || hl.toUpperCase();               // fallback: "fr" → "FR"

  if (!ucid) {
    ucid = await getChannelUCID();
  }
  if (!ucid) {
    console.error(`${window.YoutubeAntiTranslate.LOG_PREFIX} could not find channel UCID`);
    return;
  }

  // build the request body
  const body = {
    context: {
      client: {
        hl, // language
        gl, // region
        clientName: "WEB",
        clientVersion: "2.20250527.00.00"
      }
    },
    browseId: ucid,
    // “about” tab protobuf — old first, then fall back to the newer encoding
    params: "EgVhYm91dA=="      // {2:string:"about"} :contentReference[oaicite:0]{index=0}
  };

  const requestIdentifier = `youtubei/v1/browse_${body}`

  // Check cache
  if (window.YoutubeAntiTranslate.cache.has(requestIdentifier)) {
    return window.YoutubeAntiTranslate.cache.get(requestIdentifier);
  }

  const ep = "https://www.youtube.com/youtubei/v1/browse?prettyPrint=false";
  let res = await fetch(ep, { method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify(body) });

  // YT sometimes rejects the legacy protobuf with 400; try the newer one.
  if (res.status === 400) {
    body.params = "EgVhYm91dPIGBAoCEgA=";              // wrapped “about” protobuf (2023-present) :contentReference[oaicite:1]{index=1}
    res = await fetch(ep, { method: "POST", headers: { "content-type": "application/json" }, body: JSON.stringify(body) });
  }

  const json = await res.json();

  const hdr = json.header?.pageHeaderRenderer;
  const metadata = json.metadata?.channelMetadataRenderer;

  const result = {
    title: metadata?.title, // channel name
    truncatedDerscription: hdr?.content?.pageHeaderViewModel?.description?.descriptionPreviewViewModel?.description?.content,
    description: metadata?.description // full description
  };

  console.log(window.YoutubeAntiTranslate.LOG_PREFIX, result);

  // Store in cache
  window.YoutubeAntiTranslate.cache.set(requestIdentifier, result);

  return result;
}

// EXAMPLE ─ get branding in Italian for current channel
getChannelBranding(null, "it-IT");
```

zpix1 commented:
Added you as collaborator. Please don't merge yet without my approval. Also please modify CI, so 1. It should pass 2. If env is not correct, it should exit with error, not hang like that.

zpix1 commented:
About branding, still think that any solution without api key is better than solution with api key. 

We can add api key as another mode, but it should work for all features (replace hacky ones)

namakeingo commented:
> About branding, still think that any solution without api key is better than solution with api key. 
> 
> We can add api key as another mode, but it should work for all features (replace hacky ones)

I understand your point of view. I will change channelBrandig to use youtubei+i18n by default (if no api key is provided) and fallback to YouTube Data API if the user provides an API key. 
I will also update the popup to reflect that

namakeingo commented:
>  please modify CI, so 1. It should pass 2. If env is not correct, it should exit with error, not hang like that.

that is done

> Added you as a collaborator.

I still do not have enough permission to add the secrets, tho

zpix1 commented:
Looks like it is not possible in github - https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/store-information-in-variables#creating-configuration-variables-for-a-repository

zpix1 commented:
I decided use my own secrets, debugging now

zpix1 commented:
yep, I don't use 2fa, it works locally, for now I investigate why it can't set locale remotely

namakeingo commented:
and also, do not add the account to your phone. Phone notification takes priority over OTP so the login will ask for that instead of 2fa

zpix1 commented:
I see, thanks, will do it now

zpix1 commented:
Can you please find whats the issue here: https://github.com/zpix1/yt-anti-translate/actions/runs/15391494930
